### PR TITLE
Manipulating the DOMs with Refs 1/2

### PR DESCRIPTION
**Pull request by imnotmoon**
asdf

### CODE REVIEW

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import {Code} from "../../../src/components/Code";React는 여러분의 컴포넌트가 자주 수정될 필요가 없도록 렌더링 결과물에 알맞게 자동으로 [DOM](https://developer.mozilla.org/docs/Web/API/Document_Object_Model/Introduction)을 업데이트합니다.
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]: 제목/개행이 날아간느낌..?

```suggestion
import {Code} from "../../../src/components/Code";

React는 여러분의 컴포넌트가 자주 수정될 필요가 없도록 렌더링 결과물에 알맞게 자동으로 [DOM](https://developer.mozilla.org/docs/Web/API/Document_Object_Model/Introduction)을 업데이트합니다.
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+하지만 가끔은 React에 의해 관리되고 있는 DOM에 직접 접근해야 할 필요가 있습니다.
+노드를 focus 한다거나, 그 노드로 스크롤을 해야 한다거나, 사이즈와 위치를 알아야 할 때가 적절한 예시입니다.
+React에 그런 작업을 위한 빌트인 도구가 없기 때문에, _ref_를 통해 DOM 노드에 접근해야 합니다.
+
+:::info 여러분이 배울 것
+- React가 관리하는 DOM 노드에 `ref` 속성을 통해 접근하는 방법
+- 어떻게 JSX `ref` 속성이 `useRef` hook과 관계를 맺는지.
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
- JSX `ref` 속성이 `useRef` hook과 관계를 맺는 법
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+```
+
+### 예시 : 텍스트 input을 focus 하기
+이 예시에서, 버튼을 클릭하면 input이 focus 됩니다.
+
+<Code src="https://codesandbox.io/embed/sweet-banzai-k2uec3?fontsize=14&hidenavigation=1&theme=dark" />
+
+이를 구현하기 위해서는
+1. `useRef` hook을 통해 `inputRef`를 선언합니다.
+2. 이를 `<input ref={inputRef}>`에 넘깁니다. 이는 React에게 *이 `<input>`의 DOM 노드를 `inputRef.current`에 넣으라는 말과 같습니다.*
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
**이 `<input>`의 DOM 노드를 `inputRef.current`에 넣으라는 말과 같습니다.**
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+Ref에 대한 내용을 더 알고 싶다면 [ref의 참조 변수](https://beta.reactjs.org/learn/referencing-values-with-refs)를 읽어보세요.
+
+### 예시 : 요소로 스크롤 하기
+
+컴포넌트 하나에 한 개 이상의 ref를 선언할 수 있습니다. 아래 예시에 세 이미지를 가진 캐러셀이 있습니다.
+각 버튼은 브라우저의 `[scrollIntoView()](https://codesandbox.io/embed/priceless-lake-tzej7b?fontsize=14&hidenavigation=1&theme=dark)` 메소드를 불러 이에 상응하는 DOM 노드가 중간으로 오도록 합니다.
+
+<Code src="https://codesandbox.io/embed/priceless-lake-tzej7b?fontsize=14&hidenavigation=1&theme=dark" />
+
+:::note DEEP DIVE
```
Comment (ChipmunkForLove) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx:67]: How to manage a list of refs using a ref callback deep dive 제목이 빠졌어요

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+위 예시에서, 이미 정의된 ref가 몇개 있습니다. 하지만, 가끔 배열의 각 아이템 갯수만큼 ref가 필요할 수도 있습니다.
+그리고 얼마나 필요한지 여러분은 모르는 상황이죠. 아래와 같은 코드는 *동작하지 않을 겁니다.*
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
그리고 얼마나 필요한지 여러분은 모르는 상황이죠. 아래와 같은 코드는 **동작하지 않을 겁니다.**
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  {items.map((item) => {
+    // 동작하지 않음!
+    const ref = useRef(null);
+    return <li ref={ref} />;
+  })}
+</ul>
+```
+
+이는 *hook이 항상 컴포넌트의 최상단에서 호출되어야 하기 때문입니다.* `useRef`는 루프나, 조건문이나, `map()` 내부에서 호출될 수 없습니다.
+
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
이는 **hook이 항상 컴포넌트의 최상단에서 호출되어야 하기 때문입니다.** `useRef`는 루프나, 조건문이나, `map()` 내부에서 호출될 수 없습니다.
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+한가지 가능한 방법은 부모 요소에 하나의 ref를 두고, `[querySelectorAll](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll)`과 같은 메소드를 이용해 각각의 자식 노드들을 "찾는"것입니다.
+하지만 이 방법은 안정성이 떨어지고 DOM 구조가 바뀔 경우 동작하지 않을 수 있습니다.
+
+다른 방법은 *함수에 `ref` 속성을 넘기는 것*입니다. 이는 [`ref` 콜백](https://beta.reactjs.org/reference/react-dom/components/common#ref-callback)이라고 불리는 이 방법은 DOM 노드에 ref가 세팅되는 시점에 React로 하여금 콜백을 호출하도록 합니다.
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx:84]:
```suggestion
이는 **hook이 항상 컴포넌트의 최상단에서 호출되어야 하기 때문입니다.** `useRef`는 루프나, 조건문이나, `map()` 내부에서 호출될 수 없습니다.
```

Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
다른 방법은 **함수에 `ref` 속성을 넘기는 것**입니다. 이는 [`ref` 콜백](https://beta.reactjs.org/reference/react-dom/components/common#ref-callback)이라고 불리는 이 방법은 DOM 노드에 ref가 세팅되는 시점에 React로 하여금 콜백을 호출하도록 합니다.
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+```
+이 방법은 나중에 Map에서 각각의 DOM 노드를 읽도록 해줍니다.
+:::
+
+## 다른 컴포넌트의 DOM 노드 접근하기
+
+`<input>`과 같은 브라우저 요소를 출력하는 내장 구성 요소에 ref를 넣으면 React는 해당 ref의 `current` 속성을 해당 DOM 노드(브라우저의 실제 `<input />`같은)로 설정합니다.
+
+하지만, ref를 `<MyInput />`과 같이 *직접 만든* 컴포넌트에 넘겨준다면 기본값은 `null`이 될 것입니다.
+아래는 이를 보여주는 예시이빈다. 버튼을 클릭해도 input이 focus되지 *않는*다는 점에 유의하세요.
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
하지만, ref를 `<MyInput />`과 같이 **직접 만든** 컴포넌트에 넘겨준다면 기본값은 `null`이 될 것입니다.
아래는 이를 보여주는 예시이빈다. 버튼을 클릭해도 input이 focus되지 **않는**다는 점에 유의하세요.
```

Comment (ChipmunkForLove) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
아래는 이를 보여주는 예시입니다. 버튼을 클릭해도 input이 focus되지 *않는*다는 점에 유의하세요.
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+:::danger
+// console
+Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
+:::
+
+이 문제는 기본적으로 React에서 컴포넌트가 다른 컴포넌트 요소의 DOM 노드에 접근할 수 없기 때문에 발생합니다.
+심지어 그 자식 컴포넌트에도 말입니다! 이건 의도된 것입니다. ref는 조심스럽게 사용해야 하는 탈출구와 같습니다.
+다른 컴포넌트의 DOM 노드를 수동으로 조작하면 코드가 훨씬 취약해질 수 있습니다.
+
+대신 DOM 노드를 _노출_하려는 컴포넌트가 해당 동작을 *선택*해야 합니다.
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
대신 DOM 노드를 _노출_하려는 컴포넌트가 해당 동작을 **선택**해야 합니다.
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+```jsx
+const MyInput = forwardRef((props, ref) => {
+  return <input {...props} ref={ref} />;
+});
+```
+
+이 방법이 동작하는 방식은 아래와 같습니다.
+
+1. `MyInput ref={inputRef} />`는 React에게 해당 DOM 노드를 `inputRef.current`에 넣으라고 지시합니다. 하지만, 이를 선택하는 것은 `MyInput`에게 달려 있습니다. 기본적으로는 그렇지 않습니다.
+2. `MyInput`컴포넌트는 `fowradRef`를 이용해 선언되었습니다. *이를 통해 `inputRef`를 `props` 뒤의 두번째 인자인 `ref`를 통해 받을 것이라 선택합니다.*
```
Comment (Songycs) [inline docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]:
```suggestion
2. `MyInput`컴포넌트는 `fowradRef`를 이용해 선언되었습니다. **이를 통해 `inputRef`를 `props` 뒤의 두번째 인자인 `ref`를 통해 받을 것이라 선택합니다.**
```

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx (hunk 1 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]: 엇 또이러네

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx (hunk 8 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]: [aa238d5](https://github.com/LINE-NEXT-LEVEL/client/pull/32/commits/aa238d5a485dc82cb8e592c1fe19be2f2f10538e)

outdate 돼서 바로 반영이 안되네요. 따로 커밋 추가했습니다.

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx (hunk 4 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx:67]: [aa238d5](https://github.com/LINE-NEXT-LEVEL/client/pull/32/commits/aa238d5a485dc82cb8e592c1fe19be2f2f10538e)

여기서 추가했습니다.

File: docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx (hunk 3 above)
Comment (imnotmoon) [reply docs/learn/Escape Hatches/manipulating-the-doms-with-refs.mdx]: [aa238d5](https://github.com/LINE-NEXT-LEVEL/client/pull/32/commits/aa238d5a485dc82cb8e592c1fe19be2f2f10538e)

commit suggestion하면 앞 내용까지 날아가서 별도로 작성했습니다.

**Review by Songycs — APPROVED** [review]
👍 👍 👍

**Comment by Songycs** [comment]
선머지필씅

**Pull request merged**